esp-hal/src/gpio/mod.rs

```diff
@@ -252,68 +253,34 @@ pub trait RtcPinWithResistors: RtcPin {
 }
 
 /// Common trait implemented by pins
-pub trait Pin: private::Sealed {
+pub trait Pin: Sealed {
     /// GPIO number
     fn number(&self, _: private::Internal) -> u8;
 
-    /// Type-erase (degrade) this pin into an ErasedPin.
+    /// Type-erase (degrade) this pin into an AnyPin.
     ///
     /// This converts pin singletons (`GpioPin<0>`, …), which are all different
     /// types, into the same type. It is useful for creating arrays of pins,
     /// or avoiding generics.
-    fn degrade(self) -> ErasedPin
+    fn degrade(self) -> AnyPin
     where
         Self: Sized,
     {
         self.degrade_internal(private::Internal)
     }
 
     #[doc(hidden)]
-    fn degrade_internal(&self, _: private::Internal) -> ErasedPin;
+    fn degrade_internal(&self, _: private::Internal) -> AnyPin;
 
     /// Enable/disable sleep-mode
     fn sleep_mode(&mut self, on: bool, _: private::Internal);
 
     /// Configure the alternate function
     fn set_alternate_function(&mut self, alternate: AlternateFunction, _: private::Internal);
-
-    /// Listen for interrupts
-    fn listen(&mut self, event: Event, _: private::Internal) {
-        self.listen_with_options(event, true, false, false, private::Internal)
-    }
-
-    /// Checks if listening for interrupts is enabled for this Pin
-    fn is_listening(&self, _: private::Internal) -> bool {
-        is_listening(self.number(private::Internal))
-    }
-
-    /// Listen for interrupts
-    fn listen_with_options(
-        &mut self,
-        event: Event,
-        int_enable: bool,
-        nmi_enable: bool,
-        wake_up_from_light_sleep: bool,
-        _: private::Internal,
-    );
-
-    /// Stop listening for interrupts
-    fn unlisten(&mut self, _: private::Internal);
-
-    /// Checks if the interrupt status bit for this Pin is set
-    fn is_interrupt_set(&self, _: private::Internal) -> bool;
-
-    /// Clear the interrupt status bit for this Pin
-    fn clear_interrupt(&mut self, _: private::Internal);
-
-    /// Enable this pin as a wake up source
-    fn wakeup_enable(&mut self, enable: bool, event: WakeEvent, _: private::Internal) {
-        self.listen_with_options(event.into(), false, false, enable, private::Internal);
-    }
 }
 
-/// Trait implemented by pins which can be used as inputs
-pub trait InputPin: Pin {
+/// Trait implemented by pins which can be used as peripheral inputs.
+pub trait PeripheralInputPin: Sealed {
```

Comment:
My only note on the name is the `Pin` suffix could probably be dropped, since it's not always a pin.

Reply:
Just for the record, I'm not perfectly happy with this abstraction, regardless of the suffix (will fix!). Peripheral drivers use these traits to configure the GPIO pads themselves, which is both convenient, and very annoying at the same time. Convenient, because as a user I don't usually care to do this config, but as a third party driver author I might want to :)

Peripheral I/O shouldn't, in an ideal world, care about GPIO drive strength, pull resistors, input/output mode, and this leaky abstraction really shows in the random null methods we must have on DummyPin/Level now.